mitmproxy/addons/dns_resolver.py

```diff
@@ -155,7 +38,85 @@ async def dns_request(self, flow: dns.DNSFlow) -> None:
         )
         if should_resolve:
-            # TODO: We need to handle overly long responses here.
-            flow.response = await resolve_message(
-                flow.request, asyncio.get_running_loop()
+            all_ip_lookups = (
+                flow.request.query
+                and flow.request.op_code == dns.op_codes.QUERY
+                and all(
+                    q.type in (dns.types.A, dns.types.AAAA)
+                    and q.class_ == dns.classes.IN
+                    for q in flow.request.questions
+                )
+            )
+            # For ip_lookups (A/AAAA only) that need to be looked up in the hosts file first we use
+            # `mitmproxy_rs.DnsResolver`and for other query types we forward it to the specified name server.
+            if all_ip_lookups and ctx.options.use_hosts_file:
+                # TODO: We need to handle overly long responses here.
+                flow.response = await self.resolve_message(flow.request)
+            else:
+                flow.server_conn.address = (ctx.options.name_servers[0], 53)
+
+    async def resolve_message(self, message: dns.Message) -> dns.Message:
+        try:
+            rrs: list[dns.ResourceRecord] = []
+            for question in message.questions:
+                rrs.extend(await self.resolve_question(question))
+        except ResolveError as e:
+            return message.fail(e.response_code)
+        else:
+            return message.succeed(rrs)
+
+    async def resolve_question(
+        self, question: dns.Question
+    ) -> Iterable[dns.ResourceRecord]:
+        assert question.type in (dns.types.A, dns.types.AAAA)
+
+        try:
+            if question.type == dns.types.A:
+                addrinfos = await self.resolver.lookup_ipv4(question.name)
+            elif question.type == dns.types.AAAA:
+                addrinfos = await self.resolver.lookup_ipv6(question.name)
+        except socket.gaierror as e:
+            if e.args[0] == "NXDOMAIN":
+                raise ResolveError(dns.response_codes.NXDOMAIN)
+            elif e.args[0] == "NOERROR":
+                raise ResolveError(dns.response_codes.NOERROR)
+            else:
+                raise ResolveError(dns.response_codes.SERVFAIL)
+
+        return map(
+            lambda addrinfo: dns.ResourceRecord(
+                name=question.name,
+                type=question.type,
+                class_=question.class_,
+                ttl=dns.ResourceRecord.DEFAULT_TTL,
+                data=ipaddress.ip_address(addrinfo).packed,
+            ),
+            addrinfos,
+        )
+
+    def load(self, loader):
+        loader.add_option(
+            "use_hosts_file",
+            bool,
+            True,
+            "Use the hosts file for DNS lookups in regular DNS mode/wireguard mode.",
+        )
+
+        loader.add_option(
+            "name_servers",
+            Sequence[str],
+            mitmproxy_rs.get_system_dns_servers(),
```

Comment:
Let's default this to empty list and then do `ctx.options.name_servers or mitmproxy_rs.get_system_dns_servers()` below. Two reasons:

1. get_system_dns_servers() may potentially raise.
2. The config in `_options_gen.ts` shouldn't be platform specific.

Reply:
Do you think we should implement a `Sequence[str] | None` option type for this case?